File: examples/parameter_estimators/example_DREM_estimator.py
#!/usr/bin/env python3

import sdu_estimators

import numpy as np
import matplotlib.pyplot as plt

def main():
    dt = 0.01
    tend = int(50 / dt)

    gamma = np.array([[10.], [10.]]).flatten() * 2e6

    ell = 10.
    r = 1.

    theta_init = np.zeros((2, 1)).flatten()
    theta_true = np.array([[1.], [2.]]).flatten()

    intg_method = sdu_estimators.integrator.RK4

    reg_ext = sdu_estimators.regressor_extensions.Kreisselmeier_1x2(dt, ell, intg_method)

    solver = sdu_estimators.parameter_estimators.DREM(dt, gamma, theta_init, reg_ext, r, intg_method)

    all_theta_est = np.zeros((tend, 2))

    tvec = []

    for i in range(tend):
        t = i * dt
        tvec.append(t)
        phi = np.array([[np.sin(t)], [np.cos(t)]])
        y = phi.T @ theta_true

        solver.step(y.flatten(), phi.flatten())

        all_theta_est[i, :] = solver.get_estimate()

    print(all_theta_est)

    plt.figure()
    plt.plot(tvec, all_theta_est, label="Estimated")
    plt.hlines(theta_true, tvec[0], tvec[-1], linestyles='dashed', label='Actual')
    plt.legend()
    plt.grid()

    plt.show()


if __name__ == "__main__":
    main()

File: examples/parameter_estimators/example_DREM_estimator.cpp
#include <Eigen/Dense>
#include <chrono>
#include <cmath>
#include <fstream>
#include <iostream>
#include <vector>

#include "sdu_estimators/parameter_estimators/drem.hpp"
#include "sdu_estimators/regressor_extensions/kreisselmeier.hpp"
#include "sdu_estimators/integrator/integrator.hpp"

using namespace sdu_estimators;

int main()
{
  float dt = 0.001;
  float tend = 50 / dt; // 10s

  Eigen::Matrix<double, 2, 1> gamma;
  gamma << 10,
           10;

  gamma *= 1;

  Eigen::Matrix<double, 2, 1> theta_init, theta_true;
  // theta_init.resize(2);
  // theta_true.resize(2);

  theta_init << 0,
                0;
  theta_true << 1,
                2;

  float ell = 0.95;
  sdu_estimators::regressor_extensions::Kreisselmeier<double, 1, 2> reg_ext(dt, ell);

  // Eigen::Vector<double, 2> alpha, beta;
  // alpha << 5, 10;
  // beta << 0.1, 0.5;
  // sdu_estimators::regressor_extensions::LTI<double, 1, 2> reg_ext(dt, alpha, beta);

  // std::cout << "test" << std::endl;
  // std::vector<int> d{0, 100};
  // sdu_estimators::regressor_extensions::Delay<double, 1, 2> reg_ext(d);

  std::cout << "test" << std::endl;

  float r = 1;
  integrator::IntegrationMethod intg_method = integrator::IntegrationMethod::RK4;
  sdu_estimators::parameter_estimators::DREM<double, 1, 2> DREM(dt, gamma, theta_init, &reg_ext, r, intg_method);
  // sdu_estimators::parameter_estimators::GradientEstimator grad_est(dt, gamma, theta_init);
  std::vector<Eigen::VectorXd> all_theta_est;
  Eigen::VectorXd y;
  Eigen::VectorXd phi;
  y.resize(1);
  phi.resize(2);

  float t;

  std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();

  for (int i = 0; i < tend; ++i)
  {
    t = i * dt;
    phi << std::sin(t),
           std::cos(t);
    y << phi.transpose() * theta_true;

    DREM.step(y, phi);
    Eigen::VectorXd tmp = DREM.get_estimate();

    // save data
    all_theta_est.push_back(tmp);
    // std::cout << tmp.transpose() << std::endl;
  }

  std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();

  // std::cout << "Time difference = " << std::chrono::duration_cast<std::chrono::seconds>(end - begin).count() << "[s]" << std::endl;
  std::cout << "Time difference = " << std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count() << "[ms]" << std::endl;

  // Write all_theta_est to file
  std::ofstream outfile;
  outfile.open ("data_DREM.csv");

  outfile << "timestamp,theta_est_1,theta_est_2,theta_act_1,theta_act_2" << std::endl;

  for (int i = 0; i < tend; ++i)
  {
    outfile << i * dt << "," << all_theta_est[i][0] << "," << all_theta_est[i][1]
            << "," << theta_true[0] << "," << theta_true[1] << std::endl;
  }

  outfile.close();
}
